include/revng/TupleTree/TrackingVisits.h

```diff
@@ -0,0 +1,239 @@
+
+      collectImpl(LHSElement);
+      Stack.pop_back();
+    }
+  }
+
+  template<NotTupleTreeCompatible T>
+  void collectImpl(const T &LHS) {}
+};
+
+template<typename M>
+struct TrackingClearer {
+
+  void run(const M &LHS) { clearImpl(LHS); }
+
+private:
+  template<size_t I = 0, typename T>
+  void clearTuple(const T &LHS) {
+    if constexpr (I < std::tuple_size_v<T>) {
+
+      clearImpl(LHS.template untrackedGet<I>());
+      LHS.template clearTracker<I>();
+
+      // Recur
+      clearTuple<I + 1>(LHS);
+    }
+  }
+
+  template<StrictSpecializationOf<UpcastablePointer> T>
+  void clearImpl(const T &LHS) {
+    LHS.upcast([&](auto &LHSUpcasted) { clearImpl(LHSUpcasted); });
+  }
+
+  template<TupleSizeCompatible T>
+  void clearImpl(const T &LHS) {
+    clearTuple(LHS);
+  }
+
+  template<revng::detail::SetOrKOC T>
+  void clearImpl(const T &LHS) {
+    for (auto &LHSElement : LHS.Content) {
+      clearImpl(LHSElement);
+    }
+    LHS.clearTracking();
+  }
+
+  template<NotTupleTreeCompatible T>
+  void clearImpl(const T &LHS) {}
+};
+
+template<typename M>
+struct TrackingPusher {
+
+  void run(const M &LHS) { pushImpl(LHS); }
+
+private:
+  template<size_t I = 0, typename T>
+  void pushTuple(const T &LHS) {
+    if constexpr (I < std::tuple_size_v<T>) {
+
+      LHS.template pushTracker<I>();
+      pushImpl(LHS.template untrackedGet<I>());
+
+      // Recur
+      pushTuple<I + 1>(LHS);
+    }
+  }
+
+  template<StrictSpecializationOf<UpcastablePointer> T>
+  void pushImpl(T &LHS) {
+    LHS.upcast([&](const auto &LHSUpcasted) { pushImpl(LHSUpcasted); });
+  }
+
+  template<TupleSizeCompatible T>
+  void pushImpl(const T &LHS) {
+    pushTuple(LHS);
+  }
+
+  template<revng::detail::SetOrKOC T>
+  void pushImpl(const T &LHS) {
+    LHS.trackingPush();
+    for (auto &LHSElement : LHS.Content) {
+      pushImpl(LHSElement);
+    }
+  }
+
+  template<NotTupleTreeCompatible T>
+  void pushImpl(const T &LHS) {}
+};
+
+template<typename M>
+struct TrackingPopper {
+
+  void run(const M &LHS) { popImpl(LHS); }
+
+private:
+  template<size_t I = 0, typename T>
+  void popTuple(const T &LHS) {
+    if constexpr (I < std::tuple_size_v<T>) {
+
+      popImpl(LHS.template untrackedGet<I>());
+      LHS.template popTracker<I>();
+
+      // Recur
+      popTuple<I + 1>(LHS);
+    }
+  }
+
+  template<StrictSpecializationOf<UpcastablePointer> T>
+  void popImpl(T &LHS) {
+    LHS.upcast([&](const auto &LHSUpcasted) { popImpl(LHSUpcasted); });
+  }
+
+  template<TupleSizeCompatible T>
+  void popImpl(const T &LHS) {
+    popTuple(LHS);
+  }
+
+  template<revng::detail::SetOrKOC T>
+  void popImpl(const T &LHS) {
+    for (auto &LHSElement : LHS.Content) {
+      popImpl(LHSElement);
+    }
+    LHS.trackingPop();
+  }
+
+  template<NotTupleTreeCompatible T>
+  void popImpl(const T &LHS) {}
+};
+
+} // namespace revng::detail
+
+template<TupleSizeCompatible T>
+ReadFields collectReadFields(const T &Object) {
```

Comment:
Can we promote this to methods?

This header contains quite a bit of private stuff, but these functions are very much to be used by users.
I feel like users should ignore the existence of `TrackingVisits.h` and just use some very simple methods on the `struct`s themselves.

Reply:
i tried this, but it can't be done with the current hierarchy, to be a method it must be a method of the user defined class, not of the base generated class, otherwise you fail to be able to tell apart if the current generated class is TupleTreeCompatible. Anyway it should not really matter, there will be just one user of this ideally